include/lock/optimistic_lock.hpp

```diff
@@ -0,0 +1,233 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CPP_UTILITY_OPTIMISTIC_LOCK_HPP
+#define CPP_UTILITY_OPTIMISTIC_LOCK_HPP
+
+#ifndef SPINLOCK_HINT
+#ifdef CPP_UTILITY_HAS_SPINLOCK_HINT
+#include <xmmintrin.h>
+#define SPINLOCK_HINT _mm_pause();  // NOLINT
+#else
+#define SPINLOCK_HINT /* do nothing */
+#endif
+#endif
+
+#include <xmmintrin.h>
+
+#include <atomic>
+#include <chrono>
+#include <thread>
+#include <utility>
+
+namespace dbgroup::lock
+{
+class OptimisticLock
+{
+ public:
+  /*####################################################################################
+   * Public constructors and assignment operators
+   *##################################################################################*/
+
+  constexpr OptimisticLock() = default;
+
+  OptimisticLock(const OptimisticLock&) = delete;
+  OptimisticLock(OptimisticLock&&) = delete;
+
+  auto operator=(const OptimisticLock&) -> OptimisticLock& = delete;
+  auto operator=(OptimisticLock&&) -> OptimisticLock& = delete;
+
+  /*####################################################################################
+   * Public destructors
+   *##################################################################################*/
+
+  ~OptimisticLock() = default;
+
+  /*####################################################################################
+   * Public getters
+   *##################################################################################*/
+
+  /**
+   * @brief Get the Version object
+   *
+   * @return std::pair<uint64_t, bool>
+   */
+  [[nodiscard]] auto
+  GetVersion()  //
+      -> std::pair<uint64_t, bool>
+  {
+    version_ = lock_.load(std::memory_order_relaxed);
+    if (version_ & (2UL) == 0b10) return std::make_pair(version_, false);
+    return std::make_pair(version_, true);
+  }
+
+  /**
+   * @brief Check whether the version after reading is equal to the version before reading
+   *
+   * @return true
+   * @return false
+   */
+  auto
+  CheckVersion()  //
+      -> bool
+  {
+    std::pair<uint64_t, bool> current_version_ = GetVersion();
+    if (!GetVersion().second) return false;
+    if (current_version_.first == version_) return true;
+    return false;
+  }
+
+  /**
+   * @brief Get an exclusive lock.
+   *
+   * NOTE: this function does not give up acquiring locks and continues spinlock.
+   */
+  void
+  LockX()
+  {
+    auto expected = lock_.load(std::memory_order_relaxed);
+    auto desired = expected + kXLock;
+
+    while (!lock_.compare_exchange_weak(expected, desired, std::memory_order_acquire,
+                                        std::memory_order_relaxed)) {
+      desired = expected + kXLock;
+      SPINLOCK_HINT
+    }
```

Comment:
排他ロックを取得する際は他のスレッドがX/SIXロックを取っていないことを保証する必要があるので，XLockMaskで期待する値を修正してあげる必要がありますね．また，リトライが多発した場合は一旦waitしたほうがよいので以下のような感じでお願いします．

```suggestion
    while (true) {
      auto expected = lock_.load(std::memory_order_relaxed) & kXLockMask;
      auto desired = expected + kXLock;
      for (size_t i = 0; i < kRetryNum; ++i) {
        const auto cas_success = lock_.compare_exchange_weak(
            expected, desired, std::memory_order_acquire, std::memory_order_relaxed);
        if (cas_success) return;

        expected &= kXLockMask;
        desired = expected + kXLock;
        SPINLOCK_HINT
      }

      std::this_thread::sleep_for(kShortSleep);
    }
```